dash/_grouping.py

```diff
@@ -0,0 +1,234 @@
+
+    Note: Scalar values in schema are not used
+
+    :param schema: Grouping value encoding the structure of the grouping to return
+    :param flat_values: List of values with length matching the grouping_len of schema.
+        Elements of flat_values will become the scalar values in the resulting grouping
+    """
+
+    def _perform_make_grouping_like(value, next_values):
+        if isinstance(value, (tuple, list)):
+            return list(
+                _perform_make_grouping_like(el, next_values)
+                for i, el in enumerate(value)
+            )
+
+        if isinstance(value, dict):
+            return {
+                k: _perform_make_grouping_like(v, next_values)
+                for i, (k, v) in enumerate(value.items())
+            }
+
+        return next_values.pop(0)
+
+    if not isinstance(flat_values, list):
+        raise ValueError(
+            "The flat_values argument must be a list. "
+            "Received value of type {typ}".format(typ=type(flat_values))
+        )
+
+    expected_length = len(flatten_grouping(schema))
+    if len(flat_values) != expected_length:
+        raise ValueError(
+            "The specified grouping pattern requires {n} elements but received {m}\n"
+            "    Grouping patter: {pattern}\n"
+            "    Values: {flat_values}".format(
+                n=expected_length,
+                m=len(flat_values),
+                pattern=repr(schema),
+                flat_values=flat_values,
+            )
+        )
+
+    return _perform_make_grouping_like(schema, list(flat_values))
+
+
+def map_grouping(fn, grouping):
+    """
+    Map a function over all of the scalar values of a grouping, maintaining the
+    grouping structure
+
+    :param fn: Single-argument function that accepts and returns scalar grouping values
+    :param grouping: The grouping to map the function over
+    :return: A new grouping with the same structure as input grouping with scalar
+        values updated by the input function.
+    """
+    if isinstance(grouping, (tuple, list)):
+        return [map_grouping(fn, g) for g in grouping]
+
+    if isinstance(grouping, dict):
+        return {k: map_grouping(fn, g) for k, g in grouping.items()}
+
+    return fn(grouping)
+
+
+def make_grouping_by_key(schema, source, default=None):
+    """
+    Create a grouping from a schema by ujsing the schema's scalar values to look up
+    items in the provided source object.
+
+    :param schema: A grouping of potential keys in source
+    :param source: Dict-like object to use to look up scalar grouping value using
+        scalar grouping values as keys
+    :param default: Default scalar value to use if grouping scalar key is not present
+        in source
+    :return: grouping
+    """
+    return map_grouping(lambda s: source.get(s, default), schema)
+
+
+class SchemaValidationError(InvalidCallbackReturnValue):
+    def __init__(self, msg):
+        super().__init__(msg=msg)
```

Comment:
Does this `__init__` change anything? Unless I'm missing something we'll still go through `DashException.__init__` with or without this, and that one dedents and strips the message - so the explicit 4-space indentation you give all the messages below will be removed, and we should be able to turn each collection of single-line-with-terminator f-strings into one big triple-quoted f-string, using whatever indentation is convenient since it'll be removed anyway.

Not a big deal, just a little cleaner IMO :)

Reply:
Yeah, good idea. done in 449f16e. I also removed the `SchemaValidationError` base class because it's not actually doing anything